doc/designs/ech-api.md

```diff
@@ -0,0 +1,528 @@
+input value provided contains no suitable ECHConfig values (e.g. if it only
+contains ECHConfig versions that are not supported), then these functions will
+fail and return zero.
+
+```c
+int SSL_ech_set1_echconfig(SSL *s, const unsigned char *val, size_t len);
+int SSL_CTX_ech_set1_echconfig(SSL_CTX *ctx, const unsigned char *val,
+                               size_t len);
+```
+
+ECHConfig values may be provided via a command line argument to the calling
+application or (more likely) have been retrieved from DNS resource records by
+the application. ECHConfig values may be provided in various encodings (base64,
+ascii hex or binary) each of which may suit different applications.  ECHConfig
+values may also be provided embedded in the DNS wire encoding of HTTPS or SVCB
+resource records or in the equivalent zone file presentation format.
+
+``ossl_ech_find_echconfigs()`` attempts to find and return the (possibly empty)
+set of ECHConfig values from a buffer containing one of the encoded forms
+described above. Each successfully returned ECHConfigList will have
+exactly one ECHConfig, i.e., a single public value.
+
+```c
+int ossl_ech_find_echconfigs(int *num_echs,
+                             unsigned char ***echconfigs, size_t **echlens,
+                             const unsigned char *val, size_t len);
+```
+
+``ossl_ech_find_echconfigs()`` returns the number of ECHConfig values from the
+input (``val``/``len``) successfully decoded  in the ``num_echs`` output.  If
+no ECHConfig values values are encountered (which can happen for good HTTPS RR
+values) then ``num_echs`` will be zero but the function returns 1. If the
+input contains more than one (syntactically correct) ECHConfig, then only
+those that contain locally supported options (e.g. AEAD ciphers) will be
+returned. If no ECHConfig found has supported options then none will be
+returned and the function will return 0.
+
+After a call to ``ossl_ech_find_echconfigs()``, the application can make a
+sequence of calls to ``SSL_ech_set1_echconfig()`` for each of the ECHConfig
+values found.  (The various output buffers must be freed by the client
+afterwards, see the example code in
+[``test/ech_test.c``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/test/ech_test.c).)
+
+Clients can additionally more directly control the values to be used for inner
+and outer SNI and ALPN values via specific APIs. This allows a client to
+override the ``public_name`` present in an ECHConfigList that will otherwise
+be used for the outer SNI. The ``no_outer`` input allows a client to emit an
+outer CH with no SNI at all.
+
+```c
+int SSL_ech_set_server_names(SSL *s, const char *inner_name,
+                             const char *outer_name, int no_outer);
+int SSL_ech_set_outer_server_name(SSL *s, const char *outer_name, int no_outer);
+int SSL_CTX_ech_set_outer_alpn_protos(SSL *s, const unsigned char *protos,
+                                      unsigned int protos_len);
+```
```

Comment:
Why is the "protos" list for the "CTX" only and not a specific handle?

Reply:
I didn't find client code the needed that so far. I used to have such an API but removed it as it wasn't used. It'd be no problem to add back if needed.

Reply:
My gut feeling on the client end is that ECH is fundamentally per connection, much more than it is per context, so I'd expect the full set of functionality at the `SSL` connection object, and then, optionally, some support at the `SSL_CTX` object to the extent that there are good reasons to expect the relevant settings to be used across many connections, justifying a dedicated `SSL_CTX` with those settings, or that the settings in fact apply across all servers...